[interactive-widget] Implement logic to process interactive-widgets values

jesxilin opened the pull request:
#### ea3e2c0755e17fc3b98082a54b21347cebad5448
<pre>
[interactive-widget] Implement logic to process interactive-widgets values
<a href="https://bugs.webkit.org/show_bug.cgi?id=296721">https://bugs.webkit.org/show_bug.cgi?id=296721</a>
<a href="https://rdar.apple.com/157158598">rdar://157158598</a>

Reviewed by Abrar Rahman Protyasha and Aditya Keerthi.

Add plumbing (from web process to ui process)
for the viewport meta tag interactive-widgets values,
which consist of:
1. resize-visual
2. resize-content
3. overlays-content

Set default value of interactive-widget to resizes-visual,
which is the current default behavior on iOS.

Alter setViewportFeature to include logic for checking the
feature flag.

* Source/WebCore/dom/ViewportArguments.cpp:
(WebCore::ViewportArguments::resolve const):
(WebCore::parseInteractiveWidgetValue):
(WebCore::setViewportFeature):
* Source/WebCore/dom/ViewportArguments.h:
(WebCore::ViewportArguments::ViewportArguments):
(WebCore::ViewportArguments::operator== const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
(WebKit::RemoteLayerTreeTransaction::viewportMetaTagInteractiveWidget const):
(WebKit::RemoteLayerTreeTransaction::setViewportMetaTagInteractiveWidget):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _didCommitLayerTree:]):
(-[WKWebView _clearOverrideLayoutParameters]):
(viewportArgumentsFromDictionary):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::willCommitLayerTree):

Canonical link: <a href="https://commits.webkit.org/298096@main">https://commits.webkit.org/298096@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5288db699bc0eefb5b8e0fcd56b527c04594aac6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114236 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33983 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24445 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120402 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/64970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/21c9407d-dfd0-4952-bc0c-da1b3e26c77c) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116125 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34612 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42544 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/86816 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/64970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7cc476a1-453d-4ea3-bbe9-e4f20b5b79a6) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117184 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27562 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102598 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67206 "Passed tests") | | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20724 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64095 "Built successfully") | | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20841 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123618 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41254 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/30761 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/95649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41630 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98800 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/95432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40569 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18397 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18306 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41134 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46644 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40738 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44044 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42489 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
```diff
@@ -230,6 +230,9 @@ class RemoteLayerTreeTransaction : public CanMakeCheckedPtr<RemoteLayerTreeTrans
     void setInitialScaleFactor(double scale) { m_initialScaleFactor = scale; }
 
+    WebCore::InteractiveWidget viewportMetaTagInteractiveWidget() const { return m_viewportMetaTagInteractiveWidget; }
+    void setviewportMetaTagInteractiveWidget(WebCore::InteractiveWidget interactiveWidgetValue) { m_viewportMetaTagInteractiveWidget = interactiveWidgetValue; }
```
```suggestion
    void setViewportMetaTagInteractiveWidget(WebCore::InteractiveWidget interactiveWidgetValue) { m_viewportMetaTagInteractiveWidget = interactiveWidgetValue; }
```

jesxilin replied:
Good catch, will do!

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -4843,6 +4843,7 @@ static bool isEmpty(PDFSelection *selection)
     parameters.initialScaleIsSet = true;
     parameters.shouldHonorMinimumEffectiveDeviceWidthFromClient = false;
     parameters.minimumScaleDoesNotAdaptToContent = true;
+    parameters.interactiveWidget = WebCore::InteractiveWidget::ResizesVisual;
```
This is unnecessary since it's the default value.

pxlcoder replied:
Same comment applies elsewhere.

jesxilin replied:
Will remove!

pxlcoder commented on Source/WebCore/dom/ViewportArguments.cpp:
```diff
@@ -395,6 +410,8 @@ void setViewportFeature(ViewportArguments& arguments, StringView key, StringView
         arguments.shrinkToFit = findBooleanValue(key, value, internalErrorHandler);
     else if (equalLettersIgnoringASCIICase(key, "viewport-fit"_s))
         arguments.viewportFit = parseViewportFitValue(key, value, internalErrorHandler);
+    else if (equalLettersIgnoringASCIICase(key, "interactive-widget"_s))
```
Just realized this change does not consult the feature flag. None of this logic should run when the feature is off.

You can:
- Add a `const Document&` parameter to this method, and pass it in from the caller
- Then, check `document.settings().metaViewportInteractiveWidgetEnabled()` here

jesxilin committed:
[interactive-widget] Implement logic to process interactive-widgets values
https://bugs.webkit.org/show_bug.cgi?id=296721
rdar://157158598

Reviewed by Abrar Rahman Protyasha and Aditya Keerthi.

Add plumbing (from web process to ui process)
for the viewport meta tag interactive-widgets values,
which consist of:
1. resize-visual
2. resize-content
3. overlays-content

Set default value of interactive-widget to resizes-visual,
which is the current default behavior on iOS.

Alter setViewportFeature to include logic for checking the
feature flag.

* Source/WebCore/dom/ViewportArguments.cpp:
(WebCore::ViewportArguments::resolve const):
(WebCore::parseInteractiveWidgetValue):
(WebCore::setViewportFeature):
* Source/WebCore/dom/ViewportArguments.h:
(WebCore::ViewportArguments::ViewportArguments):
(WebCore::ViewportArguments::operator== const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
(WebKit::RemoteLayerTreeTransaction::viewportMetaTagInteractiveWidget const):
(WebKit::RemoteLayerTreeTransaction::setViewportMetaTagInteractiveWidget):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _didCommitLayerTree:]):
(-[WKWebView _clearOverrideLayoutParameters]):
(viewportArgumentsFromDictionary):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::willCommitLayerTree):

Canonical link: https://commits.webkit.org/298096@main
